candidate-agreement/src/bft/accumulator.rs

```diff
@@ -0,0 +1,586 @@
+						good = true;
+					}
+				}
+			}
+		}
+
+		good
+	}
+}
+
+/// Type alias to represent a justification specifically for a prepare.
+pub type PrepareJustification<D, S> = Justification<D, S>;
+
+/// The round's state, based on imported messages.
+#[derive(PartialEq, Eq, Debug)]
+pub enum State<C, D, S> {
+	/// No proposal yet.
+	Begin,
+	/// Proposal received.
+	Proposed(C),
+	/// Seen 2f + 1 prepares for this digest.
+	Prepared(PrepareJustification<D, S>),
+	/// Seen 2f + 1 commits for a digest.
+	Committed(Justification<D, S>),
+	/// Seen 2f + 1 round-advancement messages.
+	Advanced(Option<PrepareJustification<D, S>>),
+}
+
+/// Accumulates messages for a given round of BFT consensus.
+#[derive(Debug)]
+pub struct Accumulator<C, D, V, S>
+	where
+	C: Eq + Clone,
+	D: Hash + Eq + Clone,
+	V: Hash + Eq,
+	S: Eq + Clone,
+{
+	round_number: usize,
+	threshold: usize,
+	round_proposer: V,
+	proposal: Option<C>,
+	prepares: HashMap<V, (D, S)>,
+	commits: HashMap<V, (D, S)>,
+	vote_counts: HashMap<D, (usize, usize)>,
+	advance_round: HashSet<V>,
+	state: State<C, D, S>,
+}
+
+impl<C, D, V, S> Accumulator<C, D, V, S>
+	where
+	C: Eq + Clone,
+	D: Hash + Eq + Clone,
+	V: Hash + Eq,
+	S: Eq + Clone,
+{
+	/// Create a new state accumulator.
+	pub fn new(round_number: usize, threshold: usize, round_proposer: V) -> Self {
+		Accumulator {
+			round_number,
+			threshold,
+			round_proposer,
+			proposal: None,
+			prepares: HashMap::new(),
+			commits: HashMap::new(),
+			vote_counts: HashMap::new(),
+			advance_round: HashSet::new(),
+			state: State::Begin,
+		}
+	}
+
+	/// How advance votes we have seen.
+	pub fn advance_votes(&self) -> usize {
+		self.advance_round.len()
+	}
+
+	/// Get the round number.
+	pub fn round_number(&self) -> usize {
+		self.round_number.clone()
+	}
+
+	/// Get the round proposer.
+	pub fn round_proposer(&self) -> &V {
+		&self.round_proposer
+	}
+
+	pub fn proposal(&self) -> Option<&C> {
+		self.proposal.as_ref()
+	}
+
+	/// Inspect the current consensus state.
+	pub fn state(&self) -> &State<C, D, S> {
+		&self.state
+	}
+
+	/// Import a message. Importing duplicates is fine, but the signature
+	/// and authorization should have already been checked.
+	pub fn import_message(
+		&mut self,
+		message: LocalizedMessage<C, D, V, S>,
+	)
+	{
+		// old message.
+		if message.message.round_number() != self.round_number {
+			return;
+		}
+
+		let (sender, signature) = (message.sender, message.signature);
+
+		match message.message {
+			Message::Propose(_, p) => self.import_proposal(p, sender),
+			Message::Prepare(_, d) => self.import_prepare(d, sender, signature),
+			Message::Commit(_, d) => self.import_commit(d, sender, signature),
+			Message::AdvanceRound(_) => self.import_advance_round(sender),
+		}
+	}
+
+	fn import_proposal(
+		&mut self,
+		proposal: C,
+		sender: V,
+	) {
+		if sender != self.round_proposer || self.proposal.is_some() { return }
```

Comment:
is the second clause not misbehaviour if the proposal is different to the one we already received?

Reply:
yeah, double proposing is misbehavior. misbehavior accumulation is not yet implemented for the BFT agreement portion. is that a blocker?